### PR TITLE
K shot testing

### PR DESCRIPTION
**Pull request by ChristophHoenes**
* k shot testing script
* use k shot testing as additional best model selection method

### CODE REVIEW

File: util.py [hunk 1]
```diff
@@ -149,6 +149,19 @@ def make_dataloader(dataset_id, input_ids, labels, attention_masks, batch_size=1
                             )
     return dataloader
 
+def split_episode(batch, ratio=.5):
```
Comment (baleato) [inline util.py]: Could this be no longer needed? I don't see where it's being used.

File: train_prototype.py [hunk 2]
```diff
@@ -116,6 +121,23 @@ def train(tasks, model, args, device):
                 iter_loss, acc))
             running_loss = 0.0
 
+        # evaluate in k shot fashion
+        if iterations % args.eval_every == 0:
+            test_model.load_state_dict(model.state_dict())
+            test_mean, _ = k_shot_testing(test_model, episodes, test_task, device, num_test_batches=args.num_test_batches)
```
Comment (baleato) [inline train_prototype.py]: I think that might be helpful to log this in tensorboard too. What do you think about keeping track of both test_mean and test_std for all models?

File: train_prototype.py [hunk 3]
```diff
@@ -140,6 +162,7 @@ def train(tasks, model, args, device):
     for key, value in vars(args).items():
         print(key + ' : ' + str(value))
     device = get_pytorch_device(args)
+    print(device)
```
Comment (baleato) [inline train_prototype.py:167]: I think this line was to check that your laptop was using the GPU.

File: k_shot_testing.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            W, b = model.calculate_output_params(prototypes.detach())
+            model.initialize_classifier(W, b)
+        elif zero_init and isinstance(model, MultiTaskLearner):
+            task_module_name = 'task_{}'.format(test_task.get_name())
+            out_MTL_layer = model._modules[task_module_name]
+            out_MTL_layer.weight.data = torch.zeros_like(out_MTL_layer.weight.data)
+            out_MTL_layer.bias.data = torch.zeros_like(out_MTL_layer.bias.data)
+
+        # fine-tune model with some updates on the provided episode
+        for update in range(num_updates):
```
Comment (baleato) [inline k_shot_testing.py:45]: Should we move this loop outside the episodes one. It seems we might be going full force towards solving this episode by using it `num_updates` times, this would be the same as to increasing the learning rate and doing only one pass, no?

File: util.py (hunk 1 above)
Comment (ChristophHoenes) [reply util.py]: yes can be deleted! I think it was already in an earlier version

File: train_prototype.py (hunk 2 above)
Comment (ChristophHoenes) [reply train_prototype.py]: Totally makes sense we should do

File: train_prototype.py (hunk 3 above)
Comment (ChristophHoenes) [reply train_prototype.py:167]: We leave that for now I guess and can delete it later on once the device problem is fixed

File: job_save.sh [hunk 5]
```diff
@@ -17,17 +17,21 @@ module load cuDNN/7.0.5-CUDA-9.0.176
...
 
 pip install --user -r requirements.txt
 
-srun python train.py --save_path ~/results/
+srun python train.py --save_path ~/results/${SAVE_NAME}
 
-cp ~/results/snap* ${LISA_HOME}
-cp ~/results/best* ${LISA_HOME}
-cp ~/results/runs/* ${LISA_HOME}
+cp ~/results/${SAVE_NAME}snap* ${LISA_HOME}/${SAVE_NAME}
```
Comment (baleato) [inline job_save.sh:34]: Both `~` and `pwd` point to the home directory of the user running the process, and using the `--save_path` argument should be enough to store the products (snapshots and logs) in your home directory. I think these copy lines are redundant.

File: train_prototype.py [hunk 6]
```diff
@@ -116,6 +120,26 @@ def train(tasks, model, args, device):
                 iter_loss, acc))
             running_loss = 0.0
 
+        # evaluate in k shot fashion
+        if iterations % args.eval_every == 0:
+            test_model.load_state_dict(model.state_dict())
+            test_mean, test_std = k_shot_testing(test_model, episodes, test_task, device, num_test_batches=args.num_test_batches)
+            writer.add_scalar('TestTask/Acc', test_mean)
+            writer.add_scalar('TestTask/STD', test_std)
```
Comment (baleato) [inline train_prototype.py]: Should we also pass the iterations as the third parameter? So we know after how many iterations we get a particular value. I'm unsure how does it work without it, does it take the current time?
`writer.add_scalar('TestTask/Acc', test_mean, iterations)`

File: job_save.sh (hunk 5 above)
Comment (ChristophHoenes) [reply job_save.sh:34]: I think the `~` in `--save_path` refers to the scratch directory, right?
If you do not copy it back to your home directory  (`${LISA_HOME}/${SAVE_NAME}`) it would stay on scratch or not?

File: train_prototype.py (hunk 6 above)
Comment (ChristophHoenes) [reply train_prototype.py]: No you are right, it should be as you wrote above!
Fixed it.

File: job_save.sh (hunk 5 above)
Comment (baleato) [reply job_save.sh:34]: In my scripts I only pass the --save_path and get the results in my home directory, no copy afterwards needed.

**Review by baleato — APPROVED** [review]
